File: samples/discovery/ios/ios/ComponentBoxUI/Sources/ComponentBoxUI/extensions/component.swift
//
// Created by Matt Ramotar on 3/27/22.
// Copyright (c) 2022 Dropbox, Inc. All rights reserved.
//


import Foundation
import SwiftUI
import componentbox

@available(iOS 14.0, *)
public extension Component {
    func inflate() -> some View {
        switch self {
            
        case let box as Component.Box:
            let ui = ComponentBoxUI.Box(
                id: box.id,
                components: box.components as? [componentbox.Component],
                modifier: box.modifier,
                horizontalArrangement: box.horizontalArrangement,
                verticalAlignment: box.verticalAlignment,
                action: box.action
            )
            
            return AnyView(ui.inflate())
            
        case let button as Component.Button:
            let ui = ComponentBoxUI.Button(id: button.id, components: button.components as? [componentbox.Component], modifier: button.modifier, isEnabled: button.isEnabled as? Bool, action: button.action, variant: button.variant)
            return AnyView(ui.inflate())
            
        case let column as Component.Column:
            let ui = ComponentBoxUI.Column(
                id: column.id,
                components: column.components as? [componentbox.Component],
                modifier: column.modifier,
                verticalArrangement: column.verticalArrangement,
                horizontalAlignment: column.horizontalAlignment,
                action: column.action,
                isLazy: column.isLazy as? Bool
            )
            
            return AnyView(ui.inflate())
            
        case let image as Component.Drawable:
            let ui = ComponentBoxUI.Image(id: image.id, name: image.name, url: image.url, contentDescription: image.contentDescription, modifier: image.modifier, alignment: image.alignment, contentScale: image.contentScale)
            return AnyView(ui.inflate())
            
        case let icon as Component.Vector:
            let ui = ComponentBoxUI.Icon(id: icon.id, name: icon.name, modifier: icon.modifier, alignment: icon.alignment, contentScale: icon.contentScale, color: icon.color)
            return AnyView(ui.inflate())
            
        case let row as Component.Row:
            let ui = ComponentBoxUI.Row(
                id: row.id,
                components: row.components as? [componentbox.Component],
                modifier: row.modifier,
                horizontalArrangement: row.horizontalArrangement,
                verticalAlignment: row.verticalAlignment,
                action: row.action,
                isLazy: row.isLazy as? Bool
            )
            
            return AnyView(ui.inflate())
            
        case let text as Component.Text:
            let ui = ComponentBoxUI.Text(id: text.id, modifier: text.modifier, text: text.text, color: text.color, textStyle: text.textStyle)
            return AnyView(ui.inflate())
            
        default:
            return AnyView(HStack { })
        }
    }
}

File: componentbox/src/commonMain/kotlin/com/dropbox/componentbox/foundation/Component.kt
package com.dropbox.componentbox.foundation

import kotlinx.serialization.Serializable

@Serializable
sealed class Component: Parcelable {

    @Parcelize
    @Serializable
    data class Box(
        val id: String,
        var components: MutableList<Component>? = null,
        var modifier: Modifier? = null,
        var horizontalArrangement: Arrangement? = null,
        var verticalAlignment: Alignment? = null,
        var action: String? = null,
    ) : Component(), Parcelable

    @Parcelize
    @Serializable
    data class Row(
        val id: String,
        var components: MutableList<Component>? = null,
        var modifier: Modifier? = null,
        var horizontalArrangement: Arrangement? = null,
        var verticalAlignment: Alignment? = null,
        var action: String? = null,
        var isLazy: Boolean? = null
    ) : Component(), Parcelable

    @Parcelize
    @Serializable
    data class Column(
        val id: String,
        var components: MutableList<Component>? = null,
        var modifier: Modifier? = null,
        var verticalArrangement: Arrangement? = null,
        var horizontalAlignment: Alignment? = null,
        var isLazy: Boolean? = null,
        var isTable: Boolean? = null,
        var action: String? = null
    ) : Component(), Parcelable

    @Parcelize
    @Serializable
    data class Text(
        val id: String,
        var modifier: Modifier? = null,
        var text: String? = null,
        var color: Color? = null,
        var textStyle: String? = null
    ) : Component(), Parcelable

    @Parcelize
    @Serializable
    data class Button(
        val id: String,
        var components: MutableList<Component>? = null,
        var modifier: Modifier? = null,
        var isEnabled: Boolean? = null,
        var action: String? = null,
        var variant: String? = null,
    ) : Component(), Parcelable

    @Parcelize
    @Serializable
    data class Switch(
        val id: String,
        var isChecked: Boolean? = null,
        var modifier: Modifier? = null,
        var action: String? = null,
    ) : Component(), Parcelable

    @Parcelize
    @Serializable
    data class Drawable(
        val id: String,
        var name: String? = null,
        var url: String? = null,
        val contentDescription: String? = null,
        var modifier: Modifier? = null,
        var alignment: Alignment? = null,
        var contentScale: ContentScale? = null
    ) : Component(), Parcelable

    @Parcelize
    @Serializable
    data class Vector(
        val id: String,
        var name: String? = null,
        var modifier: Modifier? = null,
        var alignment: Alignment? = null,
        var contentScale: ContentScale? = null,
        var color: Color? = null
    ) : Component(), Parcelable

    @Parcelize
    @Serializable
    data class Surface(
        val id: String,
        var modifier: Modifier? = null,
        var shape: Shape? = null,
        var color: Color? = null,
        var contentColor: Color? = null,
        var borderStroke: BorderStroke? = null,
        var elevation: Int? = null,
        var components: MutableList<Component>? = null,
    ) : Component(), Parcelable
}

